### ISSUE TITLE
Kullanılacak kütüphaneler ve kodlama stili

### ISSUE THREAD

**Issue by onur** [aciksozluk/altyapi-tartisma#18]
Benim önerilerim:

* [Express](http://expressjs.com/) web çatısı
* [Handlebars](http://handlebarsjs.com/) şablon motoru
* ORM yerine [node-postgres](https://github.com/brianc/node-postgres)
* [Grunt](http://gruntjs.com/) görev çalıştırıcısı
* [SASS](http://sass-lang.com/) Süper güçleri olan CSS
* [Mocha](https://mochajs.org/) test çatısı

### Kodlama stili

Bu proje çok fazla geliştirici tarafından geliştirileceğinden kodlama stilinin de önceden belirlenmesinde yarar var. Kod belgelemesi için [JSDoc](http://usejsdoc.org/) kullanılabilir. Kodlama stili olarak [airbnb](https://github.com/airbnb/javascript)'nin hazırlamış olduğu kodlama stili kullanılabilir. Öne çıkan özellikleri:

* Tab yerine boşluk.
* Satırlar `100` karakterden uzun olamaz.
* `2` boşluk indent.
* Değişkenler için camelCase kullanılması: `buBirDeğişken`
* Sınıf isimleri her zaman büyük başlar: `Sınıf`
* ...

**Comment by ygunayer** [comment]
Benim önerilerim de aslında benzer, fark olarak
- node-postgres yerine [sequelize](https://github.com/sequelize/sequelize) (ORM iyi mi kötü mü konusuna hiç girmeden ORM önereyim dedim)
- Grunt yerine [gulp](https://github.com/gulpjs/gulp)
- Assertion kütüphanesi olarak [chai](https://github.com/chaijs/chai)
- Handlebars yerine [Jade](http://jade-lang.org) (ya da yeni adıyla *pug*, şu anda pek stabil görünmediğin kabul ediyorum, ama yasal bir nedenden ötürü isim değiştirmek durumunda kalmışlar ve geçiş yeni başlamış, o yüzden kısa sürede stabilite sağlanacaktır)

Öte yandan, TODO'daki maddede bu issue'nun REST API teknolojilerini belirleme konusu olduğu belirtilmiş ama yeri gelmişken server-side rendering yapılacaksa bile (bunda hala net bir sonuca varamadık sanıyorum) bunun REST API tarafında değil, REST API'ı karşılayacak bir ara web sunucusunda yapılması gerektiğini düşündüğümü belirteyim.

**Comment by ramesaliyev** [comment]
Ben de ORM kullanmayı öneriyorum. PostgreSQL için sanırım **sequelize**'dan daha iyisi yok.

**Express** +1
**Gulp** +1 (v4 çıkacak.)
**Mocha**, **Chai** +1
**Jade** +1
**SCSS** +1
**JSDocs** +1

Kodlama stili için JSLint veya JSHint gibi bir linter kullanalım derim.
Şuradaki karşılaştırmaya bir bakalım -> http://www.sitepoint.com/comparison-javascript-linting-tools/ (vaktim olunca bakıp oyumu kullanacağım birinden yana.)

Front-End tarafına şu an çok kafa yormamayı öneriyorum. Zaten ortada bir tasarım yok. Front-End'e geçileceği zaman kütüphaneleri ekstradan tartışır gerekirse farklı frameworkler ile birkaç poc yaparız diye düşünüyorum.

Ve ek bir çılgınlık olarak **ES6** +1 (ancak çok ısrarcı olmayacağım sıcak bakılmazsa)

**Comment by KiPSOFT** [comment]
Sequelize kullanılması taraftarıyım biraz saç baş yoldurabilir ama başarılı olacaktır. Veritabanı tasarımı sequelize model dosyaları ile yapılmalı. ESLint kullanılabilir.
Backend framework seçimi ayrı bir başlıkta tartışalım. Ben express yerine daha komplike bir framework öneriyorum şahsi görüşüm total.js - kütüphanede bire bir geliştirme yapabiliriz ama meteor.js de düşünlebilinir. JSDocs yerine APIDoc bir göz gezdirilebilir.

**Comment by hqv** [comment]
Burada biraz daha contribution gerekli. Daha fazla önerisi olan?

**Comment by KiPSOFT** [comment]
Bu başlığı backend olarak ayıralım. Şu konularda bir karar lazım;

- MVC Backend framework (express.js, total.js, meteor.js, sail.js)
- Orm (Sequelize)
- Şablon motoru (Jade, Handlebars)
- Dökümantasyon (JSDoc, APIDoc)

**Comment by ramesaliyev** [comment]
@KiPSOFT backend'ten ziyade ikiye ayrılması lazım; api ve server-side-renderer diye.

total.js meteor.js sail.js gibi full-paket çözümlere karşıyım. express yeterli olacaktır.
ORM'yi destekliyorum. template engine jade ya da handlebars olabilir.
Dökümantasyon için APIDoc'u inceledikten sonra fikir belirteceğim.

**Comment by onur** [comment]
aciksozluk/aciksozluk-api#3 için, kullanılacak web framework'una biran önce karar vermemiz gerekli. 

Öncelikle ne kullanmamızdan çok neden kullanmamız gerektiğine bakalım. Bence express.js kullanmalıyız, **çünkü** expressjs sinatra benzeri bir micro framework. Öğrenmesi kolay ve inanılmaz derece esnek. Middleweare sistemi sayesinde yeni özellikler çok kolay bir şekilde modüler olarak eklenebiliyor.

Bu konuda nasıl karar arabileceğimizden tam emin değilim. Belki web framework seçimini full MVC frameworku vs micro framework olarak düşünebiliriz.

**Comment by ramesaliyev** [comment]
@onur katılıyorum, micro ve üzerine büyük mimariler kurabilmemiz için çok kullanışlı olacaktır. Full vs Micro karşılaştırmasında, micro +1.

**Comment by hqv** [comment]
Genellemek istiyorum: full-fledged frameworkler kötüdür. Micro+1

**Comment by ygunayer** [comment]
Micro +1

**Comment by hqv** [comment]
Şu anda bunları kullanan pull request'ler var:

Express web çatısı
sequelize ORM
Gulp görev çalıştırıcısı
Mocha test çatısı

SCSS ve Jade/Handlebars front-end ile ilgili. Back end kütüphaneleri ile sorunu olan var mı?

**Comment by KiPSOFT** [comment]
Bir tek dökümantasyon kararı lazım.

**Comment by cevem** [comment]
Gulp da olabilir.
